docker/util.sh

```diff
@@ -5,6 +5,42 @@
 #
 # Utility library to use in scripts
 
+# This function installs packages with specific versions and also pin them
+# (with prio 990) to those versions by appending pinning information to
+# /etc/apt/preferences.d/cachalot, so they are not upgraded accidentally.
+#
+# All packages should be specified with version: pkg=version.
+#
+# Version can have wildcards, see apt-get/apt_preferences documentation for
+# details
+apt_pin_install()
+{
+	( { set +x -e; } 2>/dev/null # dissable verboseness (and silently)
+
+	# Pin the packages
+	for pkg_ver in "$@"
+	do
+		pkg="$(echo "$pkg_ver" | cut -d= -f1)"
+		ver="$(echo "$pkg_ver" | cut -d= -f2-)"
+
+		# If we still have a version, print the pinning spec to Prio
+		# 990 means don't install newer versions, but don't downgrade
+		# if installed is newer.
+		# https://manpages.debian.org/stretch/apt/apt_preferences.5.en.html#How_APT_Interprets_Priorities
+		cat <<EOT >> /etc/apt/preferences.d/cachalot
```

Comment:
Do we want to do `ftruncate -s0 /etc/apt/preferences.d/cachalot` just in case first?

Reply:
The answer is no. Otherwise if we call this command twice it will undo what the previous command did. I think is safe to assume that upstream ubuntu won't ship a `/etc/apt/preferences.d/cachalot`, but we could do the tructation in the base image if we are really concerned about this.

Reply:
You're right, we're calling this function several times per image generation.

Reply:
Not yet in reality, but we might in the future :)